File: implementation/scipamato/core/persistence-jooq/src/test/java/ch/difty/scipamato/core/persistence/newsletter/JooqNewsletterTopicRepoTest.java
package ch.difty.scipamato.core.persistence.newsletter;

import static ch.difty.scipamato.common.TestUtilsKt.assertDegenerateSupplierParameter;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.fail;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

import java.util.Collections;

import org.jooq.DSLContext;
import org.jooq.SelectJoinStep;
import org.jooq.impl.DSL;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import ch.difty.scipamato.common.DateTimeService;
import ch.difty.scipamato.core.db.tables.records.NewsletterTopicRecord;
import ch.difty.scipamato.core.entity.newsletter.NewsletterTopicDefinition;
import ch.difty.scipamato.core.entity.newsletter.NewsletterTopicFilter;
import ch.difty.scipamato.core.persistence.OptimisticLockingException;

@ExtendWith(MockitoExtension.class)
class JooqNewsletterTopicRepoTest {

    @Mock
    private DSLContext      dslContextMock;
    @Mock
    private DateTimeService dateTimeServiceMock;

    private JooqNewsletterTopicRepo repo;

    @BeforeEach
    void setUp() {
        repo = new JooqNewsletterTopicRepo(dslContextMock, dateTimeServiceMock);
    }

    @Test
    void findingCodesOfClass_withNullLanguageId_throws() {
        assertDegenerateSupplierParameter(() -> repo.findAll(null), "languageCode");
    }

    @Test
    void insertingNewsletterTopicDefinition_withNullArgument_throws() {
        assertDegenerateSupplierParameter(() -> repo.insert(null), "entity");
    }

    @Test
    void insertingNewsletterTopicDefinition_withEntityWithNonNullId_throws() {
        NewsletterTopicDefinition ntd = new NewsletterTopicDefinition(1, "de", 1);
        try {
            repo.insert(ntd);
            fail("Should have thrown exception");
        } catch (Exception ex) {
            assertThat(ex)
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessage("id must be null.");
        }
    }

    @Test
    void updatingNewsletterTopicDefinition_withNullArgument_throws() {
        assertDegenerateSupplierParameter(() -> repo.update(null), "entity");
    }

    @Test
    void updatingNewsletterTopicDefinition_withEntityWithNullId_throws() {
        NewsletterTopicDefinition ntd = new NewsletterTopicDefinition(null, "de", 1);
        assertDegenerateSupplierParameter(() -> repo.update(ntd), "entity.id");
    }

    @Test
    void applyingWhereCondition_withNullFilter_returnsTrueCondition() {
        SelectJoinStep<NewsletterTopicRecord> selectStep = mock(SelectJoinStep.class);
        repo.applyWhereCondition(null, selectStep);
        verify(selectStep).where(DSL.trueCondition());
    }

    @Test
    void applyingWhereCondition_withFilterWithNoTitleMask_returnsTrueCondition() {
        SelectJoinStep<NewsletterTopicRecord> selectStep = mock(SelectJoinStep.class);
        NewsletterTopicFilter filter = new NewsletterTopicFilter();
        assertThat(filter.getTitleMask()).isNull();

        repo.applyWhereCondition(filter, selectStep);

        verify(selectStep).where(DSL.trueCondition());
    }

    @Test
    void handlingUpdatedRecord_withNullRecord_throwsOptimisticLockingException() {
        final NewsletterTopicDefinition entity = new NewsletterTopicDefinition(10, "de", 100);
        final int userId = 5;
        try {
            repo.handleUpdatedRecord(null, entity, userId);
            fail("should have thrown exception");
        } catch (Exception ex) {
            assertThat(ex)
                .isInstanceOf(OptimisticLockingException.class)
                .hasMessage(
                    "Record in table 'newsletter_topic' has been modified prior to the update attempt. Aborting.... [NewsletterTopicDefinition(id=10)]");
        }
    }

    @Test
    void addingOrThrowing_withNullRecord_throwsOptimisticLockingException() {
        try {
            repo.addOrThrow(null, Collections.emptyList(), "nttObject");
            fail("should have thrown exception");
        } catch (Exception ex) {
            assertThat(ex)
                .isInstanceOf(OptimisticLockingException.class)
                .hasMessage(
                    "Record in table 'newsletter_topic_tr' has been modified prior to the update attempt. Aborting.... [nttObject]");
        }
    }

    @Test
    void logingOrThrowing_withDeleteCount0_throws() {
        try {
            repo.logOrThrow(0, 10, "delObj");
            fail("should have thrown exception");
        } catch (Exception ex) {
            assertThat(ex)
                .isInstanceOf(OptimisticLockingException.class)
                .hasMessage(
                    "Record in table 'newsletter_topic' has been modified prior to the delete attempt. Aborting.... [delObj]");
        }
    }
}

File: implementation/scipamato/core/persistence-jooq/src/test/kotlin/ch/difty/scipamato/core/persistence/newsletter/JooqNewsletterTopicRepoTest.kt
package ch.difty.scipamato.core.persistence.newsletter

import ch.difty.scipamato.common.DateTimeService
import ch.difty.scipamato.core.db.tables.records.NewsletterTopicRecord
import ch.difty.scipamato.core.entity.newsletter.NewsletterTopicDefinition
import ch.difty.scipamato.core.entity.newsletter.NewsletterTopicFilter
import ch.difty.scipamato.core.persistence.OptimisticLockingException
import com.nhaarman.mockitokotlin2.mock
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.fail
import org.jooq.DSLContext
import org.jooq.SelectJoinStep
import org.jooq.impl.DSL
import org.junit.jupiter.api.Test
import org.mockito.Mockito.verify

internal class JooqNewsletterTopicRepoTest {

    private val dslContextMock = mock<DSLContext>()
    private val dateTimeServiceMock = mock<DateTimeService>()

    private var repo = JooqNewsletterTopicRepo(dslContextMock, dateTimeServiceMock)

    @Test
    fun insertingNewsletterTopicDefinition_withEntityWithNonNullId_throws() {
        val ntd = NewsletterTopicDefinition(1, "de", 1)
        try {
            repo.insert(ntd)
            fail<Any>("Should have thrown exception")
        } catch (ex: Exception) {
            assertThat(ex)
                .isInstanceOf(IllegalArgumentException::class.java)
                .hasMessage("id must be null.")
        }
    }

    @Test
    fun applyingWhereCondition_withNullFilter_returnsTrueCondition() {
        val selectStep: SelectJoinStep<NewsletterTopicRecord> = mock()
        repo.applyWhereCondition<NewsletterTopicRecord>(null, selectStep)
        verify<SelectJoinStep<NewsletterTopicRecord>>(selectStep).where(DSL.trueCondition())
    }

    @Test
    fun applyingWhereCondition_withFilterWithNoTitleMask_returnsTrueCondition() {
        val selectStep: SelectJoinStep<NewsletterTopicRecord> = mock()
        val filter = NewsletterTopicFilter()
        assertThat(filter.titleMask).isNull()

        repo.applyWhereCondition<NewsletterTopicRecord>(filter, selectStep)

        verify<SelectJoinStep<NewsletterTopicRecord>>(selectStep).where(DSL.trueCondition())
    }

    @Test
    fun handlingUpdatedRecord_withNullRecord_throwsOptimisticLockingException() {
        val entity = NewsletterTopicDefinition(10, "de", 100)
        val userId = 5
        try {
            repo.handleUpdatedRecord(null, entity, userId)
            fail<Any>("should have thrown exception")
        } catch (ex: Exception) {
            assertThat(ex)
                .isInstanceOf(OptimisticLockingException::class.java)
                .hasMessage(
                    "Record in table 'newsletter_topic' has been modified prior to the update attempt. " +
                        "Aborting.... [NewsletterTopicDefinition(id=10)]"
                )
        }
    }

    @Test
    fun addingOrThrowing_withNullRecord_throwsOptimisticLockingException() {
        try {
            repo.addOrThrow(null, emptyList(), "nttObject")
            fail<Any>("should have thrown exception")
        } catch (ex: Exception) {
            assertThat(ex)
                .isInstanceOf(OptimisticLockingException::class.java)
                .hasMessage(
                    "Record in table 'newsletter_topic_tr' has been modified prior to the update attempt. " +
                        "Aborting.... [nttObject]"
                )
        }
    }

    @Test
    fun loggingOrThrowing_withDeleteCount0_throws() {
        try {
            repo.logOrThrow(0, 10, "delObj")
            fail<Any>("should have thrown exception")
        } catch (ex: Exception) {
            assertThat(ex)
                .isInstanceOf(OptimisticLockingException::class.java)
                .hasMessage(
                    "Record in table 'newsletter_topic' has been modified prior to the delete attempt. " +
                        "Aborting.... [delObj]"
                )
        }
    }
}
